calyx-opt/src/passes/simplify_static_guards.rs

```diff
@@ -0,0 +1,141 @@
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir as ir;
+use calyx_ir::Interval;
+
+#[derive(Default)]
+/// Simplifies Static Guards
+/// In particular if g = g1 & g2 & ...gn, then it takes all of the g_i's that
+/// are "static timing intervals", e.g., %[2:3], and combines them into one
+/// timing interval.
+/// For example: (port.out | !port1.out) & (port2.out == port3.out) & %[2:8] & %[5:10] ?
+/// becomes (port.out | !port1.out) & (port2.out == port3.out) & %[5:8] ?
+/// by "combining" %[2:8] & %[5:10]
+pub struct SimplifyStaticGuards;
+
+impl Named for SimplifyStaticGuards {
+    fn name() -> &'static str {
+        "simplify-static-guards"
+    }
+
+    fn description() -> &'static str {
+        "Simplify Static Guards"
+    }
+}
+
+impl SimplifyStaticGuards {
+    /// takes in g, and separates the "anded intervals" from the rest of the guard.
+    /// In other words, if we can rewrite g as g1 & g2 & .... gn, then
+    /// we take all of the g_i's that are static timing intervals (e.g., %[2:3])
+    /// and return a vec of (u64, u64)s. We also return the Some(rest of guard) (i.e.,
+    /// the parts that aren't "anded" intervals) if they exist
+    /// e.g.:
+    /// port.out & port1.out & %[3:5] & %[4:6] -> Some(port.out & port1.out), vec[(3,5), (4,6)]
+    /// %[3:5] & %[4:6] -> None, vec[(3,5), (4,6)]
+    pub fn separate_anded_intervals(
+        g: ir::Guard<ir::StaticTiming>,
+    ) -> (Option<ir::Guard<ir::StaticTiming>>, Vec<(u64, u64)>) {
```

Comment:
This method with allocate and merge vectors from each recursive call. Is there a way to rewrite it such that everything gets accumulated in the same vector through the recursive calls?